src/role_management.c

```diff
@@ -0,0 +1,706 @@
+	unsigned i;
+};
+
+struct handover_voter_data
+{
+	struct dqlite_node *node;
+	dqlite_node_id target_id;
+	char *leader_addr;
+	dqlite_node_id leader_id;
+};
+
+static int domainCount(uint64_t needle, const struct compare_data *data)
+{
+	unsigned i;
+	for (i = 0; i < data->n; i += 1) {
+		if (data->domains[i].domain == needle) {
+			return data->domains[i].count;
+		}
+	}
+	return 0;
+}
+
+static void addDomain(uint64_t domain, struct compare_data *data)
+{
+	unsigned i;
+	for (i = 0; i < data->n; i += 1) {
+		if (data->domains[i].domain == domain) {
+			data->domains[i].count += 1;
+			return;
+		}
+	}
+	if (i < NUM_TRACKED_DOMAINS) {
+		data->domains[i].domain = domain;
+		data->domains[i].count = 1;
+		data->n += 1;
+	}
+}
+
+static void removeDomain(uint64_t domain, struct compare_data *data)
+{
+	unsigned i;
+	for (i = 0; i < data->n; i += 1) {
+		if (data->domains[i].domain == domain) {
+			if (data->domains[i].count > 0) {
+				data->domains[i].count -= 1;
+			}
+			return;
+		}
+	}
+}
+
+static int compareNodesForPromotion(const void *l, const void *r, void *p)
+{
+	struct compare_data *data = p;
+	const struct all_node_info *left = l;
+	const struct all_node_info *right = r;
+	int result;
+
+	/* Nodes whose failure domains appear fewer times are preferred. */
+	result = domainCount(left->failure_domain, data) -
+		 domainCount(right->failure_domain, data);
+	if (result != 0) {
+		return result;
+	}
+
+	/* Nodes with lower weights are preferred. */
+	result = (int)(left->weight - right->weight);
+	if (result != 0) {
+		return result;
+	}
+
+	/* We prefer to promote a standby rather than a spare. If
+	 * left->role > right->role, then right is more "senior" than left,
+	 * so we want right to come first, so return 1.*/
+	return (left->role > right->role) - (left->role < right->role);
+}
+
+static int compareNodesForDemotion(const void *l, const void *r, void *p)
+{
+	/* XXX */
+	return -compareNodesForPromotion(l, r, p);
+}
+
+static void changeCb(struct raft_change *change, int status);
+
+/* Take one role change record off the queue and apply it. */
+static void startChange(struct dqlite_node *d)
+{
+	queue *head;
+	struct change_record *rec;
+	struct raft_change *change;
+	uint64_t id;
+	int role;
+	int rv;
+
+	if (QUEUE__IS_EMPTY(&d->changes)) {
+		return;
+	}
+
+	head = QUEUE__HEAD(&d->changes);
+	QUEUE__REMOVE(head);
+	rec = QUEUE__DATA(head, struct change_record, queue);
+	id = rec->id;
+	role = rec->role;
+	raft_free(rec);
+
+	change = raft_malloc(sizeof *change);
+	if (change == NULL) {
+		return;
+	}
+	change->data = d;
+	/* TODO request ID */
+	rv = raft_assign(&d->raft, change, id, translateDqliteRole(role),
+			 changeCb);
+	if (rv != 0) {
+		/* TODO */
+		raft_free(change);
+	}
+}
+
+/* When a role change has completed, start the next one. */
+static void changeCb(struct raft_change *change, int status)
+{
+	struct dqlite_node *d = change->data;
+
+	raft_free(change);
+	if (status != 0) {
+		/* TODO */
```

Comment:
Could it be that when shutting down `status` is `RAFT_CANCELED` or other similar error but nevertheless the code starts a new change just below? That might somehow explain the `uv_close()` assertion failure.